File: task.py
import time
import json
import numpy as np




class Task:
    def __init__(self, identifier=0, size=None):
        self.identifier = identifier
        # choosee the size of the problem
        self.size = size or np.random.randint(300, 3_000)
        # Generate the input of the problem
        self.a = np.random.rand(self.size, self.size)
        self.b = np.random.rand(self.size)
        # prepare room for the results
        self.x = np.zeros((self.size))
        self.time = 0

    def work(self):
        start = time.perf_counter()
        self.x = np.linalg.solve(self.a, self.b)
        self.time = time.perf_counter() - start

    def __str__(self):
        return f"tache {self.identifier}"
    
    def to_json(self) -> str:
        info = {"identifier":self.identifier,"size":self.size,"a":self.a.tolist(),"b":self.b.tolist(),"x":self.x.tolist(),"time":self.time}
        return json.dumps(info)
        
    @staticmethod
    def from_json(text: str) -> "Task":
        data = json.loads(text)
        task = Task(identifier=data.get("identifier", 0), size=data.get("size"))
        task.time = data.get("time", 0)
        task.x = np.array(data.get("x", np.zeros((task.size,))))
        task.a = np.array(data.get("a", np.random.rand(task.size, task.size)))
        task.b = np.array(data.get("b", np.random.rand(task.size)))
        
        return task

    def __eq__(self, other: "Task") -> bool:
        if type(other) != type(self):
            return False
        return (np.array_equal(self.a, other.a) and np.array_equal(self.x, other.x)  and np.array_equal(self.b, other.b)  and self.time == other.time and self.identifier == other.identifier)
                
 
 


if __name__ == "__main__" :
     task =  Task(identifier=1)
     text = task.to_json()
     task.from_json(text)
     
    

File: task.cpp
#include <iostream>
#include <vector>
#include <random>
#include <chrono>
#include <nlohmann/json.hpp>
#include <Eigen/Dense>

using json = nlohmann::json;
using namespace Eigen;
using namespace std;

class Task {
private:
    int identifier;
    int size;
    MatrixXd a;
    VectorXd b;
    VectorXd x;
    double elapsed_time;

public:
    Task(int id = 0, int sz = -1) : identifier(id), size(sz) {
        if (size == -1) {
            random_device rd;
            mt19937 gen(rd());
            uniform_int_distribution<> dist(300, 3000);
            size = dist(gen);
        }
        a = MatrixXd::Random(size, size);
        b = VectorXd::Random(size);
        x = VectorXd::Zero(size);
        elapsed_time = 0.0;
    }

    void work() {
        auto start = chrono::high_resolution_clock::now();
        x = a.colPivHouseholderQr().solve(b);
        auto end = chrono::high_resolution_clock::now();
        elapsed_time = chrono::duration<double>(end - start).count();
    }

    string to_json() const {
        json info;
        info["identifier"] = identifier;
        info["size"] = size;
        info["a"] = vector<vector<double>>(a.rows(), vector<double>(a.cols()));
        info["b"] = vector<double>(b.data(), b.data() + b.size());
        info["x"] = vector<double>(x.data(), x.data() + x.size());
        info["time"] = elapsed_time;

        for (int i = 0; i < a.rows(); ++i)
            for (int j = 0; j < a.cols(); ++j)
                info["a"][i][j] = a(i, j);

        return info.dump();
    }

    static Task from_json(const string &text) {
        json data = json::parse(text);
        Task task(data["identifier"], data["size"]);
        task.elapsed_time = data["time"];

        for (int i = 0; i < task.size; ++i) {
            for (int j = 0; j < task.size; ++j) {
                task.a(i, j) = data["a"][i][j];
            }
            task.b(i) = data["b"][i];
            task.x(i) = data["x"][i];
        }
        return task;
    }

    bool operator==(const Task &other) const {
        return identifier == other.identifier && a.isApprox(other.a) &&
               b.isApprox(other.b) && x.isApprox(other.x) &&
               elapsed_time == other.elapsed_time;
    }

    friend ostream &operator<<(ostream &os, const Task &task) {
        os << "Task " << task.identifier;
        return os;
    }
};

int main() {
    Task task(1);
    task.work();
    string json_text = task.to_json();
    Task new_task = Task::from_json(json_text);

    cout << new_task << endl;
    return 0;
}
